File: gtsam_unstable/examples/EqVIOFilterExample.cpp
/* ----------------------------------------------------------------------------

 * GTSAM Copyright 2010, Georgia Tech Research Corporation,
 * Atlanta, Georgia 30332-0415
 * All Rights Reserved
 * Authors: Frank Dellaert, et al. (see THANKS for the full author list)

 * See LICENSE for the license information

 * -------------------------------------------------------------------------- */

/**
 * @file EqVIOFilterExample.cpp
 * @brief Run the EqVIOFilter on 10 seconds of the EuRoC MAV Vicon Room 1
 * dataset.
 * @author Rohan Bansal
 */

#include <gtsam/slam/dataset.h>
#include <gtsam_unstable/examples/EqVIOReplayHelper.h>

#include <iomanip>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>
#include <vector>

using namespace gtsam;
using namespace gtsam::eqvio;
namespace replay = gtsam::eqvio::examples;

namespace {

/**
 * @brief Hardcoded reference values used for a lightweight smoke-style replay
 * check.
 */
struct HardcodedGroundTruth {
  static Vector3 position() { return Vector3(-0.954631, -0.101702, 0.179862); }
  static Vector3 velocity() { return Vector3(-0.120739, -0.314283, 0.119599); }
};

/// Build runtime filter params from replay metadata with fallback defaults.
EqVIOFilterParams paramsFromMetadata(const replay::MetadataMap& metadata) {
  EqVIOFilterParams params;
  const auto setParam = [&metadata](double& field,
                                    std::initializer_list<const char*> keys) {
    for (const char* key : keys) {
      field = replay::metadataFiniteDouble(metadata, key, field);
    }
  };

  setParam(params.initialPointDepth, {"eqf.initial_point_depth"});
  setParam(params.initialPointVariance, {"eqf.initial_point_variance"});
  setParam(params.measurementNoiseVariance,
           {"eqf.measurement_noise_variance_norm"});
  setParam(params.outlierThresholdAbs,
           {"eqf.outlier_threshold_abs", "eqf.outlier_threshold_abs_norm"});
  setParam(params.featureRetention, {"eqf.feature_retention"});
  setParam(params.biasOmegaProcessVariance, {"eqf.process_var_bias_omega"});
  setParam(params.biasAccelProcessVariance, {"eqf.process_var_bias_accel"});
  setParam(params.attitudeProcessVariance, {"eqf.process_var_attitude"});
  setParam(params.positionProcessVariance, {"eqf.process_var_position"});
  setParam(params.velocityProcessVariance, {"eqf.process_var_velocity"});
  setParam(params.cameraAttitudeProcessVariance,
           {"eqf.process_var_cam_attitude"});
  setParam(params.cameraPositionProcessVariance,
           {"eqf.process_var_cam_position"});
  setParam(params.pointProcessVariance, {"eqf.process_var_point"});

  params.inputNoise.setZero();
  for (int idx : {0, 3, 6, 9}) {
    params.inputNoise.block<3, 3>(idx, idx).setIdentity();
  }
  const auto setInputVariance = [&metadata, &params](int idx, const char* key) {
    const double variance = replay::metadataFiniteDouble(
        metadata, key, params.inputNoise(idx, idx));
    params.inputNoise.block<3, 3>(idx, idx) *= variance;
  };
  setInputVariance(0, "eqf.input_var_gyr");
  setInputVariance(3, "eqf.input_var_acc");
  setInputVariance(6, "eqf.input_var_gyr_bias_walk");
  setInputVariance(9, "eqf.input_var_acc_bias_walk");

  return params;
}

/// Construct initial reference state using metadata-provided camera extrinsics.
State initialReferenceState(const replay::MetadataMap& metadata) {
  return State(Se23::Identity(), Bias::Identity(),
               replay::cameraExtrinsicsFromMetadata(metadata).value_or(
                   Pose3::Identity()),
               {});
}

/// Construct initial covariance from metadata overrides with sensible defaults.
Matrix initialCovarianceFromMetadata(const replay::MetadataMap& metadata,
                                     const State& xi0) {
  Matrix Sigma0 = Matrix::Identity(xi0.dim(), xi0.dim());
  const auto setInitialVarianceBlock = [&metadata, &Sigma0](
                                           int idx, const char* key,
                                           double fallback) {
    Sigma0.block<3, 3>(idx, idx) *=
        replay::metadataFiniteDouble(metadata, key, fallback);
  };

  setInitialVarianceBlock(0, "eqf.initial_var_bias_omega", 0.1);
  setInitialVarianceBlock(3, "eqf.initial_var_bias_accel", 0.1);
  setInitialVarianceBlock(6, "eqf.initial_var_attitude", 1e-4);
  setInitialVarianceBlock(9, "eqf.initial_var_position", 1e-4);
  setInitialVarianceBlock(12, "eqf.initial_var_velocity", 1e-2);
  setInitialVarianceBlock(15, "eqf.initial_var_cam_attitude", 1e-5);
  setInitialVarianceBlock(18, "eqf.initial_var_cam_position", 1e-4);
  return Sigma0;
}

/// Print compact replay statistics and terminal estimate summary.
void printSummary(const EqVIOFilterParams& params, double currentTime,
                  size_t imuCount, size_t visionFrameCount,
                  size_t visionFeatureCount, const State& estimate) {
  std::cout << "CSV replay complete.\n";
  std::cout << "Events: " << (imuCount + visionFrameCount)
            << ", IMU: " << imuCount << ", vision frames: " << visionFrameCount
            << ", vision features: " << visionFeatureCount << "\n";
  std::cout << "Measurement noise variance (normalized): "
            << params.measurementNoiseVariance << "\n";
  std::cout << std::setprecision(17);
  std::cout << "Filter time: " << currentTime << "\n";
  std::cout << std::setprecision(10);
  std::cout << "Landmarks: " << estimate.n() << "\n";
  std::cout << "Pose translation: " << estimate.pose().translation().transpose()
            << "\n";
  std::cout << "GT pose translation: "
            << HardcodedGroundTruth::position().transpose() << "\n";
  std::cout << "Velocity: " << estimate.velocity().transpose() << "\n";
  std::cout << "GT velocity: " << HardcodedGroundTruth::velocity().transpose()
            << "\n";
}

}  // namespace

/**
 * @brief Run EqVIO filter replay on bundled EuRoC-derived CSV data.
 */
int main() {
  const std::string csvPath =
      findExampleDataFile("EqVIOdata_eurocmav_room1_10sec.csv");

  try {
    const replay::MetadataMap metadata = replay::readReplayMetadata(csvPath);
    const EqVIOFilterParams params = paramsFromMetadata(metadata);
    const State xi0 = initialReferenceState(metadata);
    const Matrix Sigma0 = initialCovarianceFromMetadata(metadata, xi0);

    std::optional<EqVIOFilter> filter;
    auto camera = std::make_shared<CameraModel>(
        Pose3::Identity(), Cal3_S2(1.0, 1.0, 0.0, 0.0, 0.0));

    size_t imuCount = 0;
    size_t visionFrameCount = 0;
    size_t visionFeatureCount = 0;
    bool gravityInitialized = false;
    double currentTime = -1.0;
    std::vector<IMUInput> imuBuffer;

    replay::ReplayReader reader(csvPath);
    replay::ReplayEvent event;
    while (reader.next(event)) {
      if (event.type == replay::ReplayEvent::Type::Imu) {
        if (!filter) {
          filter.emplace(xi0, Sigma0, KeyVector{}, params);
        }
        if (!gravityInitialized) {
          filter->initializeFromIMU(event.imu);
          currentTime = event.imu.stamp;
          gravityInitialized = true;
        }

        imuBuffer.push_back(event.imu);
        ++imuCount;
        continue;
      }

      ++visionFrameCount;
      visionFeatureCount += event.vision.size();
      if (!gravityInitialized || !filter) {
        continue;
      }

      const replay::BufferedImuPropagation step =
          replay::makeBufferedImuPropagation(imuBuffer, currentTime,
                                             event.tAbs);
      if (!step.imuInputs.empty()) {
        filter->predict(step.imuInputs, step.dts);
        currentTime += step.propagatedTime;
      }
      if (step.trimCount > 0) {
        imuBuffer.erase(imuBuffer.begin(),
                        imuBuffer.begin() +
                            static_cast<std::vector<IMUInput>::difference_type>(
                                step.trimCount));
      }

      const Matrix R =
          Matrix::Identity(static_cast<int>(2 * event.vision.size()),
                           static_cast<int>(2 * event.vision.size())) *
          params.measurementNoiseVariance;
      filter->update(event.vision, camera, R);
    }

    const State finalEstimate = filter ? filter->state() : xi0;
    printSummary(params, currentTime, imuCount, visionFrameCount,
                 visionFeatureCount, finalEstimate);
  } catch (const std::exception& e) {
    std::cerr << "EqVIOFilterExample failed: " << e.what() << "\n";
    return 2;
  }

  return 0;
}

File: python/gtsam_unstable/examples/EqVIOFilterExample.py
"""
Python replay example for the unstable EqVIO filter wrapper.

This mirrors the C++ EqVIOFilterExample flow:
1. load the bundled replay CSV,
2. initialize from first IMU sample,
3. run buffered IMU propagation between vision frames,
4. apply visual updates and print a compact terminal summary.
"""

from __future__ import annotations

from dataclasses import dataclass, field
import csv
import math
from typing import Dict, List, Optional

import numpy as np
import gtsam
import gtsam_unstable
from gtsam.utils import findExampleDataFile


@dataclass
class ReplayEvent:
    kind: str
    seq: int
    t_abs: float
    frame_idx: int = -1
    imu: Optional[gtsam_unstable.eqvio.IMUInput] = None
    vision: Dict[int, np.ndarray] = field(default_factory=dict)


@dataclass
class ReplayLog:
    metadata: Dict[str, str]
    events: List[ReplayEvent]


@dataclass
class BufferedImuPropagation:
    imu_inputs: List[gtsam_unstable.eqvio.IMUInput] = field(default_factory=list)
    dts: List[float] = field(default_factory=list)
    propagated_time: float = 0.0
    trim_count: int = 0


@dataclass
class TimestampedImu:
    t_abs: float
    imu: gtsam_unstable.eqvio.IMUInput


def _parse_float(text: str, fallback: float = 0.0) -> float:
    if text is None or text == "":
        return fallback
    return float(text)


def _parse_int(text: str, fallback: int = 0) -> int:
    if text is None or text == "":
        return fallback
    return int(text)


def _metadata_finite_float(metadata: Dict[str, str], key: str, fallback: float) -> float:
    if key not in metadata or metadata[key] == "":
        return fallback
    try:
        value = float(metadata[key])
        return value if math.isfinite(value) else fallback
    except ValueError:
        return fallback


def _camera_offset_from_metadata(metadata: Dict[str, str]) -> Optional[gtsam.Pose3]:
    keys = ["T_ci_tx", "T_ci_ty", "T_ci_tz", "T_ci_qw", "T_ci_qx", "T_ci_qy", "T_ci_qz"]
    if any(k not in metadata or metadata[k] == "" for k in keys):
        return None

    tx = float(metadata["T_ci_tx"])
    ty = float(metadata["T_ci_ty"])
    tz = float(metadata["T_ci_tz"])
    qw = float(metadata["T_ci_qw"])
    qx = float(metadata["T_ci_qx"])
    qy = float(metadata["T_ci_qy"])
    qz = float(metadata["T_ci_qz"])

    norm = math.sqrt(qw * qw + qx * qx + qy * qy + qz * qz)
    if norm <= 1e-12:
        return None
    qw /= norm
    qx /= norm
    qy /= norm
    qz /= norm

    return gtsam.Pose3(gtsam.Rot3.Quaternion(qw, qx, qy, qz), np.array([tx, ty, tz]))


def _required_columns() -> List[str]:
    return [
        "row_type",
        "t_abs",
        "seq",
        "frame_idx",
        "landmark_id",
        "gx",
        "gy",
        "gz",
        "ax",
        "ay",
        "az",
        "bgx",
        "bgy",
        "bgz",
        "bax",
        "bay",
        "baz",
        "u_norm",
        "v_norm",
        "key",
        "value",
    ]


def read_replay_csv(csv_path: str) -> ReplayLog:
    metadata: Dict[str, str] = {}
    events: List[ReplayEvent] = []
    seq_to_vision_event: Dict[int, int] = {}

    with open(csv_path, newline="", encoding="utf-8") as handle:
        reader = csv.DictReader(handle)
        if reader.fieldnames is None:
            raise ValueError(f"EqVIOFilterExample: empty file: {csv_path}")

        for key in _required_columns():
            if key not in reader.fieldnames:
                raise ValueError(f"EqVIOFilterExample: missing required column: {key}")

        for row in reader:
            row_type = (row.get("row_type") or "").strip()
            if row_type == "":
                continue

            if row_type == "meta":
                key = (row.get("key") or "").strip()
                if key:
                    metadata[key] = (row.get("value") or "").strip()
                continue

            seq = _parse_int(row.get("seq"), 0)
            t_abs = _parse_float(row.get("t_abs"), 0.0)

            if row_type == "imu":
                imu = gtsam_unstable.eqvio.IMUInput(
                    t_abs,
                    np.array(
                        [
                            _parse_float(row.get("gx")),
                            _parse_float(row.get("gy")),
                            _parse_float(row.get("gz")),
                        ]
                    ),
                    np.array(
                        [
                            _parse_float(row.get("ax")),
                            _parse_float(row.get("ay")),
                            _parse_float(row.get("az")),
                        ]
                    ),
                    np.array(
                        [
                            _parse_float(row.get("bgx")),
                            _parse_float(row.get("bgy")),
                            _parse_float(row.get("bgz")),
                        ]
                    ),
                    np.array(
                        [
                            _parse_float(row.get("bax")),
                            _parse_float(row.get("bay")),
                            _parse_float(row.get("baz")),
                        ]
                    ),
                )
                events.append(ReplayEvent(kind="imu", seq=seq, t_abs=t_abs, imu=imu))
                continue

            if row_type == "vision_feature":
                frame_idx = _parse_int(row.get("frame_idx"), -1)
                landmark_id = _parse_int(row.get("landmark_id"), 0)
                u_norm = _parse_float(row.get("u_norm"), 0.0)
                v_norm = _parse_float(row.get("v_norm"), 0.0)

                if seq not in seq_to_vision_event:
                    event = ReplayEvent(
                        kind="vision",
                        seq=seq,
                        t_abs=t_abs,
                        frame_idx=frame_idx,
                        vision={landmark_id: np.array([u_norm, v_norm])},
                    )
                    seq_to_vision_event[seq] = len(events)
                    events.append(event)
                else:
                    event = events[seq_to_vision_event[seq]]
                    if event.frame_idx != frame_idx:
                        raise ValueError(
                            f"EqVIOFilterExample: inconsistent frame_idx for vision seq={seq}"
                        )
                    if abs(event.t_abs - t_abs) > 1e-9:
                        raise ValueError(
                            f"EqVIOFilterExample: inconsistent t_abs for vision seq={seq}"
                        )
                    event.vision[landmark_id] = np.array([u_norm, v_norm])
                continue

            raise ValueError(f"EqVIOFilterExample: unknown row_type: {row_type}")

    events.sort(key=lambda event: event.seq)
    return ReplayLog(metadata=metadata, events=events)


def make_buffered_imu_propagation(
    imu_buffer: List[TimestampedImu], current_time: float, target_time: float
) -> BufferedImuPropagation:
    out = BufferedImuPropagation()
    if not imu_buffer or target_time <= current_time:
        return out

    t_ref = current_time
    for i, sample in enumerate(imu_buffer):
        t0 = max(sample.t_abs, t_ref)
        t1 = min(imu_buffer[i + 1].t_abs, target_time) if i + 1 < len(imu_buffer) else target_time
        dt = max(t1 - t0, 0.0)
        if dt <= 0.0:
            continue
        out.imu_inputs.append(sample.imu)
        out.dts.append(dt)
        out.propagated_time += dt

    first_ge_target = None
    for i, sample in enumerate(imu_buffer):
        if sample.t_abs >= target_time:
            first_ge_target = i
            break
    idx = len(imu_buffer) if first_ge_target is None else first_ge_target
    if idx != 0:
        out.trim_count = idx - 1

    return out


def _initial_covariance_from_metadata(metadata: Dict[str, str]) -> np.ndarray:
    Sigma0 = np.eye(21)

    def set_initial_variance_block(idx: int, key: str, fallback: float) -> None:
        Sigma0[idx : idx + 3, idx : idx + 3] *= _metadata_finite_float(
            metadata, key, fallback
        )

    set_initial_variance_block(0, "eqf.initial_var_bias_omega", 0.1)
    set_initial_variance_block(3, "eqf.initial_var_bias_accel", 0.1)
    set_initial_variance_block(6, "eqf.initial_var_attitude", 1e-4)
    set_initial_variance_block(9, "eqf.initial_var_position", 1e-4)
    set_initial_variance_block(12, "eqf.initial_var_velocity", 1e-2)
    set_initial_variance_block(15, "eqf.initial_var_cam_attitude", 1e-5)
    set_initial_variance_block(18, "eqf.initial_var_cam_position", 1e-4)
    return Sigma0


def _configure_filter_from_metadata(
    filter_eqf: gtsam_unstable.eqvio.EqVIOFilter, metadata: Dict[str, str]
) -> float:
    initial_point_depth = _metadata_finite_float(metadata, "eqf.initial_point_depth", 10.0)
    initial_point_variance = _metadata_finite_float(
        metadata, "eqf.initial_point_variance", 1.0
    )
    measurement_noise_variance = _metadata_finite_float(
        metadata, "eqf.measurement_noise_variance_norm", 1e-4
    )
    outlier_threshold_abs = _metadata_finite_float(metadata, "eqf.outlier_threshold_abs", 1e8)
    outlier_threshold_abs = _metadata_finite_float(
        metadata, "eqf.outlier_threshold_abs_norm", outlier_threshold_abs
    )
    outlier_threshold_prob = _metadata_finite_float(
        metadata, "eqf.outlier_threshold_prob", 1e8
    )
    feature_retention = _metadata_finite_float(metadata, "eqf.feature_retention", 0.3)

    bias_omega_process_variance = _metadata_finite_float(
        metadata, "eqf.process_var_bias_omega", 0.001
    )
    bias_accel_process_variance = _metadata_finite_float(
        metadata, "eqf.process_var_bias_accel", 0.001
    )
    attitude_process_variance = _metadata_finite_float(
        metadata, "eqf.process_var_attitude", 0.001
    )
    position_process_variance = _metadata_finite_float(
        metadata, "eqf.process_var_position", 0.001
    )
    velocity_process_variance = _metadata_finite_float(
        metadata, "eqf.process_var_velocity", 0.001
    )
    camera_attitude_process_variance = _metadata_finite_float(
        metadata, "eqf.process_var_cam_attitude", 0.001
    )
    camera_position_process_variance = _metadata_finite_float(
        metadata, "eqf.process_var_cam_position", 0.001
    )
    point_process_variance = _metadata_finite_float(
        metadata, "eqf.process_var_point", 0.001
    )

    input_noise = np.zeros((12, 12))
    input_variances = {
        0: _metadata_finite_float(metadata, "eqf.input_var_gyr", 1e-3),
        3: _metadata_finite_float(metadata, "eqf.input_var_acc", 1e-3),
        6: _metadata_finite_float(metadata, "eqf.input_var_gyr_bias_walk", 1e-3),
        9: _metadata_finite_float(metadata, "eqf.input_var_acc_bias_walk", 1e-3),
    }
    for idx, variance in input_variances.items():
        input_noise[idx : idx + 3, idx : idx + 3] = np.eye(3) * variance

    filter_eqf.setInitialLandmarkParams(initial_point_depth, initial_point_variance)
    filter_eqf.setMeasurementNoiseVariance(measurement_noise_variance)
    filter_eqf.setOutlierParams(
        outlier_threshold_abs, outlier_threshold_prob, feature_retention
    )
    filter_eqf.setProcessVariances(
        bias_omega_process_variance,
        bias_accel_process_variance,
        attitude_process_variance,
        position_process_variance,
        velocity_process_variance,
        camera_attitude_process_variance,
        camera_position_process_variance,
        point_process_variance,
    )
    filter_eqf.setInputNoise(input_noise)
    return measurement_noise_variance


def main() -> None:
    csv_path = findExampleDataFile("EqVIOdata_eurocmav_room1_10sec.csv")
    log = read_replay_csv(csv_path)

    camera_offset = _camera_offset_from_metadata(log.metadata)
    initial_covariance = _initial_covariance_from_metadata(log.metadata)

    filter_eqf: Optional[gtsam_unstable.eqvio.EqVIOFilter] = None
    measurement_noise_variance = 1e-4
    imu_buffer: List[TimestampedImu] = []
    gravity_initialized = False
    current_time = -1.0

    imu_count = 0
    vision_frame_count = 0
    vision_feature_count = 0

    for event in log.events:
        if event.kind == "imu":
            if filter_eqf is None:
                filter_eqf = gtsam_unstable.eqvio.EqVIOFilter()
                measurement_noise_variance = _configure_filter_from_metadata(
                    filter_eqf, log.metadata
                )
                filter_eqf.setReferenceCovariance(initial_covariance)
                if camera_offset is not None:
                    filter_eqf.setCameraOffset(camera_offset)
            if not gravity_initialized and event.imu is not None:
                filter_eqf.initializeFromIMU(event.imu)
                current_time = event.t_abs
                gravity_initialized = True
            if event.imu is not None:
                imu_buffer.append(TimestampedImu(t_abs=event.t_abs, imu=event.imu))
                imu_count += 1
            continue

        vision_frame_count += 1
        vision_feature_count += len(event.vision)
        if not gravity_initialized or filter_eqf is None:
            continue

        step = make_buffered_imu_propagation(imu_buffer, current_time, event.t_abs)
        if step.imu_inputs:
            filter_eqf.predict(step.imu_inputs, step.dts)
            current_time += step.propagated_time
        if step.trim_count > 0:
            imu_buffer = imu_buffer[step.trim_count:]

        R = np.eye(2 * len(event.vision)) * measurement_noise_variance
        filter_eqf.update(event.vision, R)

    print("CSV replay complete.")
    print(
        f"Events: {len(log.events)}, IMU: {imu_count}, vision frames: {vision_frame_count}, "
        f"vision features: {vision_feature_count}"
    )
    print(f"Measurement noise variance (normalized): {measurement_noise_variance}")
    print(f"Filter time: {current_time:.16f}")

    if filter_eqf is None:
        return

    position = np.asarray(filter_eqf.position()).reshape(3)
    velocity = np.asarray(filter_eqf.velocity()).reshape(3)
    print(f"Landmarks: {filter_eqf.landmarkCount()}")
    print(f"Pose translation: {position[0]: .10f} {position[1]: .10f} {position[2]: .10f}")
    print("GT pose translation: -0.954631 -0.101702  0.179862")
    print(f"Velocity: {velocity[0]: .10f} {velocity[1]: .10f} {velocity[2]: .10f}")
    print("GT velocity: -0.120739 -0.314283  0.119599")


if __name__ == "__main__":
    main()
